File: dungeons-1.21.11/src/main/java/fishmod/utils/Easing.java
package fishmod.utils;

/** Cubic ease-in-out timing for smooth open/close animations. */
public final class Easing {

    private Easing() {}

    public static float easeInOutCubic(float x) {
        return x < 0.5f ? 4f * x * x * x : 1f - (float) Math.pow(-2f * x + 2f, 3) / 2f;
    }

    /** Tracks a boolean-driven 0..1 progress value, eased over {@code durationMs}. */
    public static final class Anim {
        private final long durationMs;
        private boolean target = false;
        private float startValue = 0f;
        private long startTime = 0L;

        public Anim(long durationMs) { this.durationMs = durationMs; }

        public void setTarget(boolean expand) {
            if (expand == target) return;
            startValue = progress();
            target = expand;
            startTime = System.currentTimeMillis();
        }

        public boolean target() { return target; }

        public float progress() {
            long elapsed = System.currentTimeMillis() - startTime;
            float t = durationMs <= 0 ? 1f : Math.min(1f, elapsed / (float) durationMs);
            float eased = easeInOutCubic(t);
            float end = target ? 1f : 0f;
            return startValue + (end - startValue) * eased;
        }

        public boolean isAnimating() {
            float p = progress();
            return p > 0.001f && p < 0.999f;
        }
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/utils/Easing.kt
package fishmod.utils

object Easing {

    @JvmStatic
    fun easeInOutCubic(x: Float): Float {
        return if (x < 0.5f) 4f * x * x * x else 1f - Math.pow((-2f * x + 2f).toDouble(), 3.0).toFloat() / 2f
    }

    @JvmStatic
    fun easeOutBack(x: Float): Float {
        val c1 = 1.70158f
        val c3 = c1 + 1f
        val d = x - 1f
        return 1f + c3 * d * d * d + c1 * d * d
    }

    class Anim(private val durationMs: Long) {
        private var target: Boolean = false
        private var startValue: Float = 0f
        private var startTime: Long = 0L

        fun setTarget(expand: Boolean) {
            if (expand == target) return
            startValue = progress()
            target = expand
            startTime = System.currentTimeMillis()
        }

        fun target(): Boolean = target

        fun progress(): Float {
            if (!fishmod.utils.config.values.FishSettings.fmAnimations) return if (target) 1f else 0f
            val elapsed = System.currentTimeMillis() - startTime
            val t = if (durationMs <= 0) 1f else Math.min(1f, elapsed / durationMs.toFloat())
            val eased = easeInOutCubic(t)
            val end = if (target) 1f else 0f
            return startValue + (end - startValue) * eased
        }

        fun isAnimating(): Boolean {
            val p = progress()
            return p > 0.001f && p < 0.999f
        }
    }
}
